### ISSUE TITLE
S3 redirect (301) for requesting incorrect region for bucket results in timeout instead of error.

### ISSUE THREAD

**Issue by abc2mit** [aws/aws-sdk-ios#254]
When trying to connect to an S3 bucket that is in a different region than specified, ```AWSS3TransferManager``` goes into a redirect loop, eventually failing in a timeout.

For example, say the S3 bucket is in ```AWSRegionUSWest2``` but we accidentally set the region to ```AWSRegionUSEast1```. Setting the configuration with the following will fail.

```Objective-C
AWSServiceConfiguration *serviceConfiguration = [[AWSServiceConfiguration alloc] initWithRegion:AWSRegionUSEast1 credentialsProvider:_credentialsProvider];
    [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = serviceConfiguration;
```

The response message from S3 suggests that S3 knows where the bucket is. Expected behavior is that the client will handle the 301 redirect properly, but log a warning to the console for the developer to update the configuration.

**Comment by yosuke-matsuda** [comment]
The AWS Mobile SDK for iOS requires the correct region to be set for it to work properly and does not forward requests to other regions. You need to create a service client with the correct configuration. If you do not know the location of the bucket in advance, you can use [`- getBucketLocation:`](http://docs.aws.amazon.com/AWSiOSSDK/latest/Classes/AWSS3.html#//api/name/getBucketLocation:) to get the region. Also, if you need to make requests to buckets in multiple regions, we recommend creating one service client per region.

**Comment by abc2mit** [comment]
Sure, I can do that, but I think the problem here is the disconnect between the server response and how the client handles it, such that the error is buried. The timeout error thrown by the client is a red herring and only indicates that S3 is unreachable, which is false. The server returns an ```Endpoint``` in the response, which is 1. non-standard for HTTP responses to a 301 (while technically not illegal to HTTP protocol, it does break the spirit of how a 301 should respond — with the full URI in both the Location header and in the body of the response), and 2. the client can't handle the format returned by the response, which it should.

In my case, here's an example of a curl request to the wrong S3 region.

```
$ curl -v "https://s3.amazonaws.com/tastebuddies/photos/12452C7A-99B0-412E-8AF1-64F608F602D8/2015/10/1/F54621D0-FC2D-4FF9-9950-FD4CF9C7C5D5"*   Trying 54.231.18.152...
* Connected to s3.amazonaws.com (54.231.18.152) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
* Server certificate: s3.amazonaws.com
* Server certificate: VeriSign Class 3 Secure Server CA - G3
* Server certificate: VeriSign Class 3 Public Primary Certification Authority - G5
> GET /tastebuddies/photos/12452C7A-99B0-412E-8AF1-64F608F602D8/2015/10/1/F54621D0-FC2D-4FF9-9950-FD4CF9C7C5D5 HTTP/1.1
> Host: s3.amazonaws.com
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< x-amz-request-id: 6293C3F73CE99F65
< x-amz-id-2: Jqr6mz+LEgA8F7d1LWUaoEFth1fMpoT/CqB2BJt03pFkRZU+IgcbgpSbqDoxXPQPCNlTPy3bUJc=
< Content-Type: application/xml
< Transfer-Encoding: chunked
< Date: Mon, 05 Oct 2015 21:18:28 GMT
< Server: AmazonS3
<
<?xml version="1.0" encoding="UTF-8"?>
* Connection #0 to host s3.amazonaws.com left intact
<Error><Code>PermanentRedirect</Code><Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message><Bucket>tastebuddies</Bucket><Endpoint>tastebuddies.s3.amazonaws.com</Endpoint><RequestId>6293C3F73CE99F65</RequestId><HostId>Jqr6mz+LEgA8F7d1LWUaoEFth1fMpoT/CqB2BJt03pFkRZU+IgcbgpSbqDoxXPQPCNlTPy3bUJc=</HostId></Error>
```

The problem is that there's no way to indicate to the client that I want to hit that endpoint specifically — I am required to specify a region, otherwise it won't work, but with that response, I don't know what region the bucket is in.

 So the fix requested is:
1. (required) Handle the 301 such that I, the developer, knows that it's a 301 and I can change my configuration accordingly. I should only use the ```getBucketLocation:``` if I get a 301. Pre-emptively doing that is bad because it's another network hop that will slow down my application and cost the user data.
2. (optional) Allow endpoints such as that returned by the server to actually be useable.
2. (optional) Handle the redirect appropriately such that an error message is logged and the endpoint returned by the server is used by the client.

**Comment by abc2mit** [comment]
Okay, so I'm trying to implement the ```getBucketLocation:```. How do I translate ```typedef NS_ENUM(NSInteger, AWSS3BucketLocationConstraint)``` returned by ```AWSS3GetBucketLocationOutput.locationConstraint``` to ```typedef NS_ENUM(NSInteger, AWSRegionType)``` needed by ```AWSServiceConfiguration``` ???

**Comment by yosuke-matsuda** [comment]
You cannot simply forward a request made to one region to another. Most of the requests to AWS services are signed with AWS Signature Version 4, and the region information is a part of the signature. When a request is forwarded to another region, the authentication fails. You need to know which bucket you are accessing and have the correct (and with the minimal permissions) credentials to access the bucket before making a request. In most situations, this should be handled by the application logic. Can you describe your use case so that we can understand your needs better?

You can translate `AWSS3BucketLocationConstraint` to `AWSRegionType` using `if` or `switch` statements.

**Comment by abc2mit** [comment]
I'm not expecting the server to forward a request, but again, either the client library should know how to redirect, or not handle the 301 and pass it back to me. I can then handle the redirect by finding the correct bucket. The problem isn't so much that it's an S3 issue, but that the error thrown by the server is swallowed. The client keeps trying the original URL (even though it returns a 301) until it times out.

There's no method in the library to handle the translation? I have to do it myself and maintain it if the API changes? That can be dangerous.

**Comment by yosuke-matsuda** [comment]
It will be a compiler time failure if the API changes, so it should not be a major source of bugs. (In any case, we recommend keeping track of a list of bucket names and their regions instead of trying to determine the region dynamically.) Also, the transfer manager seems to return an error right away when you have a wrong region in the service configuration. Do you have a complete code snippet to reproduce the issue?

We will phase out the transfer manager and focus our effort on the transfer utility. You should evaluate the transfer utility if it meets your requirements.

**Comment by abc2mit** [comment]
I looked at Transfer Utility. I will switch when AWSTask is supported
instead of requiring a completion block being passed in.

I no longer have a snippet, because I’m changed the setting to use the
correct region. It was a simple creation of the credentials provider and
service configuration with both pointing to region AWSRegionUSEast1. My S3
bucket was in AWSRegionUSWest2. I also was using Cognito, so I needed to
point that to AWSRegionUSEast1 instead of AWSRegionUSWest2.

The other interesting thing is that the console says the S3 bucket region
is “Oregon”, but how do I know whether it’s USWest1 or USWest2? It seems
like I can only know by trial and error. This makes bringing up new buckets
more difficult to manage.
